build_runner/lib/src/asset/reader.dart

```diff
@@ -24,9 +25,11 @@ class SinglePhaseReader implements AssetReader {
   final _globsRan = new Set<Glob>();
   final int _phaseNumber;
   final String _primaryPackage;
+  final BuildImpl _buildImpl;
```

Comment:
this dependency cycle makes me nervous.

I've always imagined the communication happening across the `AssetGraph` rather than the `BuildImpl`

Reply:
`BuildImpl` handles all the actual building, which is what we need to do (whenever somebody tries to read an output of an optional builder). Otherwise `AssetGraph` would probably have to have a `BuildImpl` which creates a similar (if not worse) situation imo.

We could instead pass in a function which lazily builds things (in this case we would just pass `this.runLazyPhaseForInput`), but that doesn't really solve the circular dependency issue... although it does restrict usage of the `BuildImpl` to that one function.

Reply:
I think that would be an improvement since it makes the communication pattern more explicit.

I think it would be nice to shift further towards an AssetGraph or related class for this type of thing but we could also revisit that later.

Reply:
Ok, done. I also changed `runLazyPhaseForInput` to private since it is no longer directly referenced by external code which is a nice side effect of this change.